components/brave_shields/adblock/rs/src/engine.rs

```diff
@@ -0,0 +1,202 @@
+}
+
+impl Engine {
+    pub fn enable_tag(&mut self, tag: &CxxString) -> EmptyTupleResult {
+        || -> Result<(), InternalError> { Ok(self.engine.enable_tags(&[tag.to_str()?])) }().into()
+    }
+
+    pub fn disable_tag(&mut self, tag: &CxxString) -> EmptyTupleResult {
+        || -> Result<(), InternalError> { Ok(self.engine.disable_tags(&[tag.to_str()?])) }().into()
+    }
+
+    pub fn tag_exists(&self, key: &CxxString) -> BoolResult {
+        || -> Result<bool, InternalError> { Ok(self.engine.tag_exists(key.to_str()?)) }().into()
+    }
+
+    pub fn matches(
+        &self,
+        url: &CxxString,
+        hostname: &CxxString,
+        source_hostname: &CxxString,
+        request_type: &CxxString,
+        third_party_request: bool,
+        previously_matched_rule: bool,
+        force_check_exceptions: bool,
+    ) -> BlockerResultResult {
+        || -> Result<BlockerResult, InternalError> {
+            let inner_blocker_result = self.engine.check_network_urls_with_hostnames_subset(
+                url.to_str()?,
+                hostname.to_str()?,
+                source_hostname.to_str()?,
+                request_type.to_str()?,
+                Some(third_party_request),
+                previously_matched_rule,
+                force_check_exceptions,
+            );
+            Ok(inner_blocker_result.into())
+        }()
+        .into()
+    }
+
+    pub fn get_csp_directives(
+        &self,
+        url: &CxxString,
+        hostname: &CxxString,
+        source_hostname: &CxxString,
+        request_type: &CxxString,
+        third_party_request: bool,
+    ) -> StringResult {
+        || -> Result<String, InternalError> {
+            Ok(self
+                .engine
+                .get_csp_directives(
+                    url.to_str()?,
+                    hostname.to_str()?,
+                    source_hostname.to_str()?,
+                    request_type.to_str()?,
+                    Some(third_party_request),
+                )
+                .unwrap_or_default())
+        }()
+        .into()
+    }
+
+    pub fn deserialize(&mut self, serialized: &CxxVector<u8>) -> EmptyTupleResult {
+        self.engine.deserialize(serialized.as_slice()).map_err(|e| InternalError::from(e)).into()
+    }
+
+    pub fn add_resource(
+        &mut self,
+        name: &CxxString,
+        content_type: &CxxString,
+        data: &CxxString,
+    ) -> EmptyTupleResult {
+        || -> Result<(), InternalError> {
+            let resource = Resource {
+                name: name.to_str().unwrap().to_string(),
+                aliases: vec![],
+                kind: ResourceType::Mime(MimeType::from(content_type.to_str()?)),
+                content: data.to_string(),
+            };
+            Ok(self.engine.add_resource(resource)?)
+        }()
+        .into()
+    }
+
+    pub fn use_resources(&mut self, resources_json: &CxxString) -> EmptyTupleResult {
```

Comment:
This is only ever provided the contents of a file from our component updater, just containing output from `JSON.stringify`. If we don't trust the serverside implementation to always work correctly (I'll try not to be offended :wink:), `bool` for success is fine here.

Reply:
Is it possible for the component updater to produce both bad UTF-8 and bad JSON (not doubting the quality of the implementation, I'm just paranoid, and i like extra checks! :smiley:)? If the answer is yes, then it could be good to reconsider for reasons [stated above](https://github.com/brave/brave-core/pull/17368/files#r1134848672). If there is only one type of error, then a boolean would be perfect, since we can easily assume the type of error when logging via `LOG(ERROR)` or other means.

Reply:
TBH if the component updater is ever somehow having errors of this nature, we'll notice _really_ fast because a bunch of adblocking (most importantly, on YouTube) is just not going to work at all. And, if that is indeed the case, we'll again be able to discover the issue very easily since the same component is shipped to everyone.

Reply:
ok, will go ahead and log a generic error